### PR TITLE
Build an apphost with hostfxr and hostpolicy linked in

### PR DESCRIPTION
**Pull request by lpereira**
This PR introduces a few changes:

- It builds most of hostfxr as a static library, with the shared library being linked with this static library and some shared-specific code
- It builds hostpolicy as a static library
- It abstracts how the host looks for hostfxr functions, either by loading a DLL and finding symbols there, or referencing symbols from the executable directly
- Provides a minimal hostpolicy_resolver that merely references some needed hostpolicy functions

I have not built this on anything other than x64 Linux.  ~~Right now, it shouldn't work anywhere else as it requires a ld-only flag to build the shared library from the static library in a way that it doesn't throw away the whole code.~~ (I've force-pushed some changes that should make the hostpolicy DLL be built correctly regardless of the OS, so there's a good chance this will work fine in all platforms.)

The generated binary, `static_apphost`, isn't currently installed anywhere. It just lives in the artifacts directory. No changes have been performed to use this when publishing an application.

One of the concerns with this was about the size of the executable. This table summarizes it:

| Executable name | Size in bytes |
|------------------|---------|
| apphost-linux-x64 | 181472 |
| static_apphost-linux-x64 | 570312 |

This work started out as a big overhaul in how the whole hosting layer was built, but as I'm leaving the .NET team and I ran out of time, I ended up restarting the work and opting for a far simpler solution.  There are a lot of things that I'd love to have done here, and I'll be happy to discuss that with anybody that wants to continue this work.

### CODE REVIEW

**Comment by lpereira** [comment]
Found an issue with the hostpolicy DLL. Will fix this tomorrow morning.

**Comment by swaroop-sridhar** [comment]
CC: @VSadov

**Comment by vitek-karas** [comment]
* Can you maybe create an issue and describe the possible improvements you had in mind?
* Minor: personally I would prefer to leave most of the source files in the `fxr`/`hostpolicy` directory and only move the cmake and specific sources to the subdirectories.
* This will need at least some tests, something super basic to validate that the thing actually runs.

File: src/installer/corehost/cli/apphost/CMakeLists.txt [hunk 1]
```diff
@@ -1,50 +1,2 @@
...
-    set_property(TARGET ${PROJECT_NAME} PROPERTY
-            LINK_FLAGS "/MANIFEST:NO"
-        )
-endif()
-
-# Specify non-default Windows libs to be used for Arm/Arm64 builds
-if (CLR_CMAKE_TARGET_WIN32 AND (CLR_CMAKE_TARGET_ARCH_ARM OR CLR_CMAKE_TARGET_ARCH_ARM64))
-    target_link_libraries(apphost Advapi32.lib shell32.lib)
-endif()
+add_subdirectory(static)
```
Comment (VSadov) [inline src/installer/corehost/cli/apphost/CMakeLists.txt:1]: License header needed?

Comment (lpereira) [reply src/installer/corehost/cli/apphost/CMakeLists.txt:5]: Done.

**Comment by VSadov** [comment]
I had issues with pushing commits to the original branch, so I will open another PR to continue with this:

https://github.com/dotnet/runtime/pull/35368

**Pull request closed without merging**